glibc: 2.40-66 -> 2.42-47

Ma27 opened the pull request:
Announcement: https://lists.gnu.org/archive/html/info-gnu/2025-01/msg00014.html

Build-wise it's looking relatively good and I got quite far.
However, I'm a little worried about

> * dlopen and dlmopen no longer make the stack executable if a shared
  library requires it, either implicitly because of a missing GNU_STACK
  ELF header (and default ABI permission having the executable bit set)
  or explicitly because of the executable bit in GNU_STACK, and the
  stack is not already executable.  Instead, loading such objects will
  fail.

Let's see how many regressions Hydra can find regarding that. But I'll need to look for more bugreports on that matter for potential fallout.

---

TODO:

* [x] Another Hydra-run with staging (there's surprisingly much stuff depending on e.g. openjdk that's fixed in upstream/staging or asyncio (that was flaky, but cannot restart only "dependency failed" builds depending on this AFAIK))
* [x] Test a full NixOS system
* [x] Add documentation for how to work around dlopen() problem in e.g. release notes
* [x] Await staging being free for 26.05


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Ma27 committed:
glibc: 2.40-66 -> 2.42-47

Announcements:
* https://lists.gnu.org/archive/html/info-gnu/2025-01/msg00014.html
* https://lists.gnu.org/archive/html/info-gnu/2025-07/msg00011.html

vcunat commented:
Jobset: https://hydra.nixos.org/jobset/nixpkgs/glibc-2.41

Linux stdenvs for now, afterwards I expect all of x86_64-linux.  Given the previous experience, I assume that you chose a base that's known to be not too broken.

vcunat commented:
A thread of interest: https://sourceware.org/bugzilla/show_bug.cgi?id=32653

fabianhjr reviewed (commented):
https://www.gamingonlinux.com/2025/02/the-glibc-2-41-update-has-been-causing-problems-for-linux-gaming/

Should probably skip this release, there is likely to be a more backwards compatible release in the near future.

trofi commented:
`dtc` will need a fix for it's test suite:
- https://github.com/dgibson/dtc/commit/ce1d8588880aecd7af264e422a16a8b33617cef7

trofi commented:
`webkitgtk` also fails the build against `glibc-2.41`. That is fixed by a `gobject-introspection` backport:
- https://github.com/NixOS/nixpkgs/pull/390512

Ma27 commented:
It was precisely the intention to not forget about this here by having a build failure when rebasing again.

Ma27 committed:
gcc13,gcc14: fix build w/ glibc-2.42

vcunat commented:
One week, only about one week left.

Ma27 committed:
autobase: remove

This has a sodium-native.node with a wrongly set executable bit.
Now, instead of patching it, I'll just drop it because it's not a real
program, but a Node library. We use `buildNpm*` for that purpose.

This only makes maintenance harder because issues like this are
non-trivial to spot!

Related: https://github.com/NixOS/nixpkgs/pull/403638#issuecomment-3339125475

Ma27 committed:
hypershell: fix build w/ glibc-2.42

Failing Hydra build: https://hydra.nixos.org/build/302187760

Sodium claims that the executable stack of this DSO is unnecessary[1],
hence we just remove it instead of working with tunables.

[1] https://github.com/holepunchto/sodium-native/issues/214

Ma27 committed:
hypercore: remove

Same reasoning as with `autobase`.

Ma27 committed:
hyper-cmd-utils: fix runtime w/ glibc-2.42

Same issue as with hypershell, `sodium-native.node` wrongly requests an
executable stack and is dlopened into a process with a non-executable
stack.

No smoke-test as install-check here since this beautiful piece of
software returns a non-zero exit code on `--help` 🤡

Ma27 committed:
hyperssh: fix runtime w/ glibc-2.42

Same issue as with hypershell and hyper-cmd-utils. Also no smoke-test
here because `--help` returns a non-zero exit code.

Ma27 committed:
corestore: remove

Same reasoning as with hypercore and autobase.

Ma27 committed:
hyperswarm: remove

Same reasoning as with hypercore and autobase.

Ma27 committed:
jdk8: fix build w/ glibc-2.42

Failing Hydra build: https://hydra.nixos.org/build/308320851

Upstream bug report: https://bugs.openjdk.org/browse/JDK-8354941
Only patching JDK8, newer versions received backports.

Ma27 committed:
llvmPackages_1{8,9}.compiler-rt: fix build w/ glibc-2.42

Failing Hydra build: https://hydra.nixos.org/build/308327424

Ma27 committed:
libpisp: fix build w/ glibc-2.42

Failing Hydra build: https://hydra.nixos.org/build/308323206

Ma27 committed:
glibc: make it easier to work with a dev shell

Creating the build directory in `..` may be OK in a sandbox, but this
makes it unnecessarily annoying when trying to work with this in a
source-checkout.

Hence, move the `build/` directory into the source tree.

Ma27 committed:
chuck: 1.4.2.0 -> 1.5.5.5, fix build w/ glibc-2.42

Failing Hydra build: https://hydra.nixos.org/build/308304933
ChangeLog: https://chuck.stanford.edu/release/VERSIONS

Ma27 committed:
xxgdb: remove

Failing Hydra build: https://hydra.nixos.org/build/308374102

Considering how old this package seems to be according to the
changelog[1], I'm not going to bother with patching it.

[1] https://metadata.ftp-master.debian.org/changelogs//main/x/xxgdb/xxgdb_1.12-17_changelog

Ma27 committed:
powwow: remove

Failing Hydra build: https://hydra.nixos.org/build/308340445

Ma27 committed:
rr: fix build w/ glibc-2.42

Failing Hydra build: https://hydra.nixos.org/build/308527220

Ma27 committed:
gccNGPackages_15: 15.1.0 -> 15.2.0, fix build w/ glibc-2.42

To fix build with glibc-2.42.
Failing Hydra build: https://hydra.nixos.org/build/308474067

Ma27 committed:
ananicy-cpp: fix build w/ glibc-2.42

Failing Hydra build: https://hydra.nixos.org/build/308467284

Ma27 committed:
rshim-user-space: mark as broken

Failing Hydra build: https://hydra.nixos.org/build/308527236

Compat is being worked on, but the patch for glibc 2.42 seems to crash
at runtime.

Ma27 committed:
rocmPackages.llvm.llvm: fix build w/ glibc-2.42

Failing Hydra build: https://hydra.nixos.org/build/308526977

I'm not sure what exactly caused the failure here, but this patch
from LLVM upstream seems to fix the problem and this regressed right
after the glibc upgrade.

Ma27 committed:
onnxruntime,python3.pkgs.onnxruntime: fix runtime w/ glibc-2.42

Failing Hydra build: https://hydra.nixos.org/build/308650242

The libraries require an execstack even though it's not necessary. This
got fixed upstream already[1], but this doesn't apply cleanly on the
version we have.

[1] https://github.com/microsoft/onnxruntime/pull/25200

Ma27 committed:
mrpeach: remove

Doesn't build with glibc-2.42 anymore, didn't see a release in 10 years.

Failing Hydra build: https://hydra.nixos.org/build/308670849

dramforever commented:
Please consider holding off this on this for a short while, at least for a short while until https://gitlab.com/qemu-project/qemu/-/issues/3065 ("Missing support for termios2 required on some architectures after glibc 2.42 upgrade") can be sorted. Otherwise it affects qemu-user use cases like non-native-arch containers and possibly non-native nix builds pretty annoyingly.

I'm going to try to handle this maybe this weekend or next week. I have a patchset almost ready, and I'll send a fix to qemu for staging-nixos if a new patch version hasn't been released by then.

Ma27 committed:
nixos/doc/release-notes: document glibc execstack change

Ma27 committed:
openvi: fix build w/ glibc-2.42

Failing Hydra build: https://hydra.nixos.org/build/317772107

Ma27 committed:
rt-tests: 2.8 -> 2.9, fix build w/ glibc-2.42

Failing Hydra build: https://hydra.nixos.org/build/317799900

This turns of a warning about a potential format overflow, but I
consider it sufficiently unlikely to have enough threads[1] such that
this actually causes an overflow.

[1] https://git.kernel.org/pub/scm/utils/rt-tests/rt-tests.git/tree/src/pmqtest/pmqtest.c#n546

Ma27 committed:
jdk11: 11.0.29+7 -> 11.0.31+0, fix build w/ glibc-2.42

Failing Hydra build: https://hydra.nixos.org/build/317760066

Ma27 commented:
OK, so I managed to rebuild my laptop's configuration on this branch and can boot into a system built with glibc 2.42.

I'd say we're good to go now. The Hydra jobset is looking quite good (and most jobs are fallout of uvloop's flaky tests anyways). There's even documentation about the execstack topic now. I'm considering to add a dedicated glibc-shenanigans section to the nixpkgs manual which will also include the execstack part, but I think we're good with the release-notes section I added.

Reviews would be very welcome.
cc @trofi @vcunat @mweinelt @k900

vcunat commented:
It sounds good to me.  Also, the executable stack thing is almost one year released upstream, so that's quite some time for other SW to adapt (at least from Nixpkgs perspective of short release cycles).

vcunat commented:
Also, more fixes could be done during `staging-next`, as usual.

fabianhjr reviewed (approved):
Only regarding glibc changes (didn't review package fixes/removals)

Would prefer removals not to be included in this PR but wouldn't oppose for this to be merged as is either and any issues being addressed during staging/staging-next cycles.
